p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,516 @@
+	DefaultKeepaliveTimeout    = 2 * time.Second
+)
+
+type WebRTCTransport struct {
+	webrtcConfig webrtc.Configuration
+	rcmgr        network.ResourceManager
+	privKey      ic.PrivKey
+	noiseTpt     *noise.Transport
+	localPeerId  peer.ID
+
+	// timeouts
+	peerConnectionTimeouts IceTimeouts
+
+	// in-flight connections
+	maxInFlightConnections uint32
+}
+
+var _ tpt.Transport = &WebRTCTransport{}
+
+type Option func(*WebRTCTransport) error
+
+type IceTimeouts struct {
+	Disconnect time.Duration
+	Failed     time.Duration
+	Keepalive  time.Duration
+}
+
+// WithPeerConnectionIceTimeouts sets the ice disconnect, failure and keepalive timeouts
+func WithPeerConnectionIceTimeouts(timeouts IceTimeouts) Option {
+	return func(t *WebRTCTransport) error {
+		if timeouts.Disconnect == 0 {
+			timeouts.Disconnect = t.peerConnectionTimeouts.Disconnect
+		}
+		if timeouts.Failed == 0 {
+			timeouts.Failed = t.peerConnectionTimeouts.Failed
+		}
+		if timeouts.Keepalive == 0 {
+			timeouts.Keepalive = t.peerConnectionTimeouts.Keepalive
+		}
+		// 0 is not treated as a default, and instead disables the respective check
+		// (eg. if keepalive interval is 0, no keepalives are sent, or if disconnected timeout is 0, the connection never enters a disconnected state).
+		// Refer here: https://github.com/pion/ice/blob/v2.3.0/agent_config.go#L73-L85
+		//
+		// The timeouts are only set to their default values if they are not set in the settingEngine.
+		// Refer here: https://github.com/pion/ice/blob/67f28cf23a8ae59a38e0128390ccdad27b5526be/agent_config.go#L213-L224
+		//
+		// Our implementation requires that the values be set (explicitly or implicitly)
+		// so that we can fail connections are free up resources if a remote disconnects abruptly,
+		// therefore 0 values are not acceptable. We also need to enforce that Failed timeout > disconnected timeout,
+		// and failed timeout > keepalive interval. (We currently enforce keepalive interval <= disconnected interval)
+		if timeouts.Disconnect != 0 {
+			if timeouts.Failed != 0 && timeouts.Failed < timeouts.Disconnect {
+				return fmt.Errorf("disconnect timeout cannot be greater than failed timeout")
+			}
+			if timeouts.Keepalive != 0 && timeouts.Disconnect <= timeouts.Keepalive {
+				return fmt.Errorf("keepalive timeout cannot be greater than or equal to failed timeout")
+			}
+		}
+		t.peerConnectionTimeouts = timeouts
+		return nil
+	}
+}
+
+// WithListenerMaxInFlightConnections sets the maximum number of connections that are in-flight, i.e
+// they are being negotiated, or are waiting to be accepted.
+func WithListenerMaxInFlightConnections(m uint32) Option {
+	return func(t *WebRTCTransport) error {
+		if m == 0 {
+			t.maxInFlightConnections = DefaultMaxInFlightConnections
+		} else {
+			t.maxInFlightConnections = m
+		}
+		return nil
+	}
+}
+
+func New(privKey ic.PrivKey, psk pnet.PSK, gater connmgr.ConnectionGater, rcmgr network.ResourceManager, opts ...Option) (*WebRTCTransport, error) {
```

Comment:
The gater is not used anywhere. It should be. lmk if you need any pointers on how to integrate it.

Reply:
Please do, as I am still not _that_ familiar with libp2p.

Reply:
For example, have a look at the WebTransport implementation: https://github.com/libp2p/go-libp2p/blob/master/p2p/transport/webtransport/transport.go.

In the case of WebRTC, it probably makes sense to also implement the `InterceptAccept` callback somewhere in / close to the muxer.

Reply:
Ok thank you. Will try to implement it this week and get back to you on it ASAP.
What is the benefit of implementing this? What do we gain with it? or why is it bad not to have it?

Reply:
libp2p users define connection gaters to block connections to / from certain nodes, either by peer ID or by IP address. Every transports needs to respect these, otherwise the connection gater wouldn't have any effect when both peers support WebRTC.

Reply:
https://github.com/libp2p/go-libp2p/pull/2200 adds a pretty comprehensive end-to-end test suite for connection gating. Once that PR is merged, you can just rebase add WebRTC, and these tests will tell you if the gater was properly integrated.

Reply:
Thank you. Merged in.